### PR TITLE
Auto-configure the client’s server URL

### PR DESCRIPTION
**Pull request by j-f1**
Corresponding PR for https://github.com/actualbudget/actual/pull/649. Also allows specifying relative paths for `webRoot` (such as `../actual/packages/desktop-client/build`, which is useful if you want to test running against a local build of the frontend).

### CODE REVIEW

**Comment by MatissJanis** [comment]
Why not use the existing `/needs-bootstrap` for this? It's used to check if the server is running actual-server or a different server after typing in the URL and clicking on "ok".

File: src/load-config.js [hunk 1]
```diff
@@ -54,7 +54,10 @@ export default {
   hostname: process.env.ACTUAL_HOSTNAME || config.hostname,
   serverFiles: process.env.ACTUAL_SERVER_FILES || config.serverFiles,
   userFiles: process.env.ACTUAL_USER_FILES || config.userFiles,
-  webRoot: process.env.ACTUAL_WEB_ROOT || config.webRoot,
+  webRoot: path.resolve(
+    projectRoot,
+    process.env.ACTUAL_WEB_ROOT || config.webRoot
+  ),
```
Comment (MatissJanis) [inline src/load-config.js:60]: **🔨 warning:** ‏the more I think about this, the more I dislike it from a security POV.

If an attacker gains access to this config file or the env variables: he basically can traverse the entire filesystem. I imagine setting it to `../../../path/to/budget.data?ignore=` and then opening up the server would spit out the budget file. 

https://www.acunetix.com/websitesecurity/directory-traversal/

**Comment by MatissJanis** [comment]
> > Why not use the existing `/needs-bootstrap` for this? It's used to check if the server is running actual-server or a different server after typing in the URL and clicking on "ok".
> 
> That’s an interesting idea! So you’re thinking we would call `/needs-bootstrap` on the current domain immediately at startup?

yup

File: src/load-config.js (hunk 1 above)
Comment (j-f1) [reply src/load-config.js:60]: If the attacker gains access to the config file then they presumably already have access to the filesystem, right?

And the environment variables are evaluated once at startup, so changes to them at runtime would not have any effect. If you were able to set them before server startup, you could already change the `ACTUAL_SERVER_FILES` environment variable to an empty directory, then re-bootstrap the app and download all of the files with the normal UI.

Comment (MatissJanis) [reply src/load-config.js:60]: Not necessarily. There might be another system running on the same instance. And the attacker might be able to upload files to it, but not read from it.

Anyways, that's a lot of _ifs_. However, we do need to consider them and try and keep the attack vector for actual-server as little as possible. Directory traversal attack is a pretty common security issue. So if we can treat **all** the input as potentially malicious - we can protect ourselves better.

In this case I really don't like the wrapping of this input in `path.resolve`. Is there an alternative we could consider that would solve your problem? Perhaps symlinks?

**Comment by j-f1** [comment]
Since d39f196b8702de79c448752b80956de0c57adabe needs more discussion, I’ll move that to a different PR and focus this on the issue at hand.

File: src/app-account.js [hunk 2]
```diff
@@ -29,7 +29,10 @@ app.get('/needs-bootstrap', (req, res) => {
 
   res.send({
     status: 'ok',
-    data: { bootstrapped: rows.length > 0 }
+    data: {
+      isActual: true,
```
Comment (MatissJanis) [inline src/app-account.js:33]: Do we really need this constant? Wouldn't a 200 response from this endpoint be sufficient to tell if it's an actual server?

Comment (j-f1) [reply src/app-account.js:33]: Sure.

**Review by MatissJanis — APPROVED** [review]

**Comment by j-f1** [comment]
Will wait to merge until the other PR is also approved to make sure there don’t need to be any other changes

**Pull request merged**